### ISSUE TITLE
Cannot access a disposed object. (Droid.PancakeDrawable)

### ISSUE THREAD

**Issue by matteopiccioni** [sthewissen/Xamarin.Forms.PancakeView#112]
Hello,
with latest version 2.0.0.691 of PancakeView on Android platform
and X.F. 4.7.0.1080
 sometimes I have the following exception

```
System.ObjectDisposedException
Cannot access a disposed object.
Object name: 'Xamarin.Forms.PancakeView.Droid.PancakeDrawable'.
```

Stack trace

>   at Java.Interop.JniPeerMembers.AssertSelf (Java.Interop.IJavaPeerable self) [0x00029] in <9324da45a6654f83baffa7c2854d836a>:0 
>   at Java.Interop.JniPeerMembers+JniInstanceMethods.InvokeNonvirtualObjectMethod (System.String encodedMember, Java.Interop.IJavaPeerable self, Java.Interop.JniArgumentValue* parameters) [0x00000] in <9324da45a6654f83baffa7c2854d836a>:0 
>   at Android.Graphics.Drawables.Drawable.get_Bounds () [0x0000a] in <eaa205f580954a64824b74a79fa87c62>:0 
>   at Xamarin.Forms.PancakeView.Droid.PancakeDrawable.PancakeViewOnPropertyChanged (System.Object sender, System.ComponentModel.PropertyChangedEventArgs e) [0x000c7] in <00f3e28af28b459fa7e4330fea9a95f7>:0 
>   at (wrapper delegate-invoke) <Module>.invoke_void_object_PropertyChangedEventArgs(object,System.ComponentModel.PropertyChangedEventArgs)
>   at Xamarin.Forms.BindableObject.OnPropertyChanged (System.String propertyName) [0x00000] in D:\a\1\s\Xamarin.Forms.Core\BindableObject.cs:229 
>   at Xamarin.Forms.Element.OnPropertyChanged (System.String propertyName) [0x00000] in D:\a\1\s\Xamarin.Forms.Core\Element.cs:353 
>   at Xamarin.Forms.PancakeView.PancakeView.InternalCollectionChanged (System.Object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e) [0x0000b] in <00f3e28af28b459fa7e4330fea9a95f7>:0 
>   at System.Collections.ObjectModel.ObservableCollection`1[T].OnCollectionChanged (System.Collections.Specialized.NotifyCollectionChangedEventArgs e) [0x00018] in /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/external/corefx/src/System.ObjectModel/src/System/Collections/ObjectModel/ObservableCollection.cs:263 
>   at Xamarin.Forms.PancakeView.GradientStopCollection.OnCollectionChanged (System.Collections.Specialized.NotifyCollectionChangedEventArgs e) [0x000b7] in <00f3e28af28b459fa7e4330fea9a95f7>:0 
>   at Xamarin.Forms.PancakeView.GradientStopCollection.CollectionItem_PropertyChanged (System.Object sender, System.ComponentModel.PropertyChangedEventArgs e) [0x00009] in <00f3e28af28b459fa7e4330fea9a95f7>:0 
>   at Xamarin.Forms.BindableObject.OnPropertyChanged (System.String propertyName) [0x00000] in D:\a\1\s\Xamarin.Forms.Core\BindableObject.cs:229 
>   at Xamarin.Forms.BindableObject.SetValueActual (Xamarin.Forms.BindableProperty property, Xamarin.Forms.BindableObject+BindablePropertyContext context, System.Object value, System.Boolean currentlyApplying, Xamarin.Forms.Internals.SetValueFlags attributes, System.Boolean silent) [0x00114] in D:\a\1\s\Xamarin.Forms.Core\BindableObject.cs:461 
>   at Xamarin.Forms.BindableObject.SetValueCore (Xamarin.Forms.BindableProperty property, System.Object value, Xamarin.Forms.Internals.SetValueFlags attributes, Xamarin.Forms.BindableObject+SetValuePrivateFlags privateAttributes) [0x00173] in D:\a\1\s\Xamarin.Forms.Core\BindableObject.cs:397 
>   at Xamarin.Forms.Internals.TypedBinding`2[TSource,TProperty].ApplyCore (System.Object sourceObject, Xamarin.Forms.BindableObject target, Xamarin.Forms.BindableProperty property, System.Boolean fromTarget) [0x0011b] in D:\a\1\s\Xamarin.Forms.Core\TypedBinding.cs:218 
>   at Xamarin.Forms.Internals.TypedBinding`2[TSource,TProperty].Apply (System.Boolean fromTarget) [0x00029] in D:\a\1\s\Xamarin.Forms.Core\TypedBinding.cs:108 
>   at Xamarin.Forms.Internals.TypedBinding`2+PropertyChangedProxy[TSource,TProperty].<OnPropertyChanged>b__16_0 () [0x00000] in D:\a\1\s\Xamarin.Forms.Core\TypedBinding.cs:277 
>   at Xamarin.Forms.DispatcherExtensions.Dispatch (Xamarin.Forms.IDispatcher dispatcher, System.Action action) [0x00028] in D:\a\1\s\Xamarin.Forms.Core\DispatcherExtensions.cs:53 
>   at Xamarin.Forms.Internals.TypedBinding`2+PropertyChangedProxy[TSource,TProperty].OnPropertyChanged (System.Object sender, System.ComponentModel.PropertyChangedEventArgs e) [0x00033] in D:\a\1\s\Xamarin.Forms.Core\TypedBinding.cs:277 
>   at Xamarin.Forms.BindingExpression+WeakPropertyChangedProxy.OnPropertyChanged (System.Object sender, System.ComponentModel.PropertyChangedEventArgs e) [0x00012] in D:\a\1\s\Xamarin.Forms.Core\BindingExpression.cs:646 
>   at (wrapper delegate-invoke) <Module>.invoke_void_object_PropertyChangedEventArgs(object,System.ComponentModel.PropertyChangedEventArgs)
>   at MvvmHelpers.ObservableObject.OnPropertyChanged (System.String propertyName) [0x00000] in /_/MvvmHelpers/ObservableObject.cs:54 
>   at MvvmHelpers.ObservableObject.SetProperty[T] (T& backingStore, T value, System.String propertyName, System.Action onChanged, System.Func`3[T1,T2,TResult] validateValue) [0x0003d] in /_/MvvmHelpers/ObservableObject.cs:39 
>   at UserApp.ViewModels.HomeViewModel.set_CentraleLive (UserApp.Model.StatoCentraleLive value) [0x00000] in /Users/.../UserApp/ViewModels/Main/HomeViewModel.cs:330 
>   at UserApp.Views.HomePage.ShowBar_Clicked (System.Object sender, System.EventArgs e) [0x00044] in /Users/.../UserApp/Views/Main/HomePage.xaml.cs:835 
>   at System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__7_0 (System.Object state) [0x00000] in /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/AsyncMethodBuilder.cs:1021 
>   at Android.App.SyncContext+<>c__DisplayClass2_0.<Post>b__0 () [0x00000] in <eaa205f580954a64824b74a79fa87c62>:0 
>   at Java.Lang.Thread+RunnableImplementor.Run () [0x00008] in <eaa205f580954a64824b74a79fa87c62>:0 
>   at Java.Lang.IRunnableInvoker.n_Run (System.IntPtr jnienv, System.IntPtr native__this) [0x00008] in <eaa205f580954a64824b74a79fa87c62>:0 
>   at (wrapper dynamic-method) Android.Runtime.DynamicMethodNameCounter.1(intptr,intptr)

**Comment by matteopiccioni** [comment]
Hello, 
The exception is in class PancakeDrawable.cs, PancakeViewOnPropertyChanged, when Bounds is disposed

```
{System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Xamarin.Forms.PancakeView.Droid.PancakeDrawable'.
  at Java.Interop.JniPeerMembers.AssertSelf (Java.Interop.IJavaPeerable self) [0x00029] in <9324da45a6654f83baffa7c2854d836a>:0 
  at Java.Interop.JniPeerMembers+JniInstanceMethods.InvokeNonvirtualObjectMethod (System.String encodedMember, Java.Interop.IJavaPeerable self, Java.Interop.JniArgumentValue* parameters) [0x00000] in <9324da45a6654f83baffa7c2854d836a>:0 
  at Android.Graphics.Drawables.Drawable.get_Bounds () [0x0000a] in <eaa205f580954a64824b74a79fa87c62>:0 
  at Xamarin.Forms.PancakeView.Droid.PancakeDrawable.PancakeViewOnPropertyChanged (System.Object sender, System.ComponentModel.PropertyChangedEventArgs e) [0x000c8] in /Users/matteo/Progetti/_GitHub/Xamarin.Forms.PancakeView/src/Xamarin.Forms.PancakeView.Multi/Platforms/Android/PancakeDrawable.cs:199 }
```

**Comment by elaurentin** [comment]
I had the same issue in Release build, Building directly from the source code works fine.

**Comment by elaurentin** [comment]
I was able to reproduce this again, found a typo in PancakeDrawable.cs (Android). Should be:
```
        protected override void Dispose(bool disposing)
        {
            if (!_isDisposed)
            {
                _pancake.PropertyChanged -= PancakeViewOnPropertyChanged;
                _isDisposed = true;

                if (disposing)
                    Reset();

                base.Dispose(disposing);
            }
        }
```
Also, minor changes in PancakeViewRender.cs (Android)
```
        protected override void OnElementChanged(ElementChangedEventArgs<PancakeView> e)
        {
            base.OnElementChanged(e);
            if (e.OldElement != null || e.NewElement == null)
            {
                // unregister handler here
                return;
            }
            ...
        }
```
```
        protected override void OnElementPropertyChanged(object sender, PropertyChangedEventArgs e)
        {
            if (_disposed) return; // very important
            base.OnElementPropertyChanged(sender, e);

            var pancake = Element as PancakeView;
            ...
        }
```

**Comment by sthewissen** [comment]
@elaurentin Thanks for that! What handler do you mean to unregister in the `OnElementChanged`? The `OnElementPropertyChanged` one?

**Comment by elaurentin** [comment]
The pancake.PropertyChanged handler, but got it in the Disposed (PancakeDrawable). I don't think that check is necessary.

**Comment by sthewissen** [comment]
Could you try this in https://www.nuget.org/packages/Xamarin.Forms.PancakeView/2.0.1.701-beta?

**Comment by elaurentin** [comment]
2.0.1.701-beta - good
If you need help, I can investigate a few other ones.

**Comment by sthewissen** [comment]
You're always allowed to do that 👍  Didn't even have to ask :)

**Comment by matteopiccioni** [comment]
Hello,
I am sorry I was away from office for a while
I try the beta version 2.0.1.701-beta, now I have exception:

```
System.MissingMethodException
Constructor on type 'Xamarin.Forms.PancakeView.Droid.PancakeViewRenderer' not found.
```
```
  at System.RuntimeType.CreateInstanceImpl (System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder binder, System.Object[] args, System.Globalization.CultureInfo culture, System.Object[] activationAttributes, System.Threading.StackCrawlMark& stackMark) [0x0022b] in /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/mcs/class/referencesource/mscorlib/system/rttype.cs:5535 
  at System.Activator.CreateInstance (System.Type type, System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder binder, System.Object[] args, System.Globalization.CultureInfo culture, System.Object[] activationAttributes) [0x000b5] in /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/mcs/class/referencesource/mscorlib/system/activator.cs:107 
  at System.Activator.CreateInstance (System.Type type, System.Object[] args) [0x00000] in /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/mcs/class/referencesource/mscorlib/system/activator.cs:112 
  at Xamarin.Forms.Internals.DependencyResolver.ResolveOrCreate (System.Type type, System.Object source, System.Type visualType, System.Object[] args) [0x00075] in D:\a\1\s\Xamarin.Forms.Core\DependencyResolver.cs:51 
  at Xamarin.Forms.Internals.Registrar`1[TRegistrable].GetHandler (System.Type type, System.Object source, Xamarin.Forms.IVisual visual, System.Object[] args) [0x00052] in D:\a\1\s\Xamarin.Forms.Core\Registrar.cs:90 
  at Xamarin.Forms.Internals.Registrar`1[TRegistrable].GetHandlerForObject[TOut] (System.Object obj, System.Object[] args) [0x0002c] in D:\a\1\s\Xamarin.Forms.Core\Registrar.cs:122 
  at Xamarin.Forms.Platform.Android.Platform.CreateRenderer (Xamarin.Forms.VisualElement element, Android.Content.Context context) [0x00000] in D:\a\1\s\Xamarin.Forms.Platform.Android\Platform.cs:347 
  at Xamarin.Forms.Platform.Android.VisualElementPackager.AddChild (Xamarin.Forms.VisualElement view, Xamarin.Forms.Platform.Android.IVisualElementRenderer oldRenderer, Xamarin.Forms.Platform.Android.RendererPool pool, System.Boolean sameChildren) [0x000b3] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:140 
  at Xamarin.Forms.Platform.Android.VisualElementPackager.SetElement (Xamarin.Forms.VisualElement oldElement, Xamarin.Forms.VisualElement newElement) [0x00163] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:351 
  at Xamarin.Forms.Platform.Android.VisualElementPackager.Load () [0x00000] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:112 
  at Xamarin.Forms.Platform.Android.VisualElementRenderer`1[TElement].SetPackager (Xamarin.Forms.Platform.Android.VisualElementPackager packager) [0x00007] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:465 
  at Xamarin.Forms.Platform.Android.VisualElementRenderer`1[TElement].SetElement (TElement element) [0x000d5] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:245 
  at Xamarin.Forms.Platform.Android.VisualElementRenderer`1[TElement].Xamarin.Forms.Platform.Android.IVisualElementRenderer.SetElement (Xamarin.Forms.VisualElement element) [0x00033] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:135 
  at Xamarin.Forms.Platform.Android.Platform.CreateRenderer (Xamarin.Forms.VisualElement element, Android.Content.Context context) [0x0001f] in D:\a\1\s\Xamarin.Forms.Platform.Android\Platform.cs:350 
  at Xamarin.Forms.Platform.Android.VisualElementPackager.AddChild (Xamarin.Forms.VisualElement view, Xamarin.Forms.Platform.Android.IVisualElementRenderer oldRenderer, Xamarin.Forms.Platform.Android.RendererPool pool, System.Boolean sameChildren) [0x000b3] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:140 
  at Xamarin.Forms.Platform.Android.VisualElementPackager.SetElement (Xamarin.Forms.VisualElement oldElement, Xamarin.Forms.VisualElement newElement) [0x00163] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:351 
  at Xamarin.Forms.Platform.Android.VisualElementPackager.Load () [0x00000] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:112 
  at Xamarin.Forms.Platform.Android.VisualElementRenderer`1[TElement].SetPackager (Xamarin.Forms.Platform.Android.VisualElementPackager packager) [0x00007] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:465 
  at Xamarin.Forms.Platform.Android.VisualElementRenderer`1[TElement].SetElement (TElement element) [0x000d5] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:245 
  at Xamarin.Forms.Platform.Android.VisualElementRenderer`1[TElement].Xamarin.Forms.Platform.Android.IVisualElementRenderer.SetElement (Xamarin.Forms.VisualElement element) [0x00033] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:135 
  at Xamarin.Forms.Platform.Android.Platform.CreateRenderer (Xamarin.Forms.VisualElement element, Android.Content.Context context) [0x0001f] in D:\a\1\s\Xamarin.Forms.Platform.Android\Platform.cs:350 
  at Xamarin.Forms.Platform.Android.VisualElementPackager.AddChild (Xamarin.Forms.VisualElement view, Xamarin.Forms.Platform.Android.IVisualElementRenderer oldRenderer, Xamarin.Forms.Platform.Android.RendererPool pool, System.Boolean sameChildren) [0x000b3] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:140 
  at Xamarin.Forms.Platform.Android.VisualElementPackager.SetElement (Xamarin.Forms.VisualElement oldElement, Xamarin.Forms.VisualElement newElement) [0x00163] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:351 
  at Xamarin.Forms.Platform.Android.VisualElementPackager.Load () [0x00000] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:112 
  at Xamarin.Forms.Platform.Android.VisualElementRenderer`1[TElement].SetPackager (Xamarin.Forms.Platform.Android.VisualElementPackager packager) [0x00007] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:465 
  at Xamarin.Forms.Platform.Android.VisualElementRenderer`1[TElement].SetElement (TElement element) [0x000d5] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:245 
  at Xamarin.Forms.Platform.Android.VisualElementRenderer`1[TElement].Xamarin.Forms.Platform.Android.IVisualElementRenderer.SetElement (Xamarin.Forms.VisualElement element) [0x00033] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:135 
  at Xamarin.Forms.Platform.Android.Platform.CreateRenderer (Xamarin.Forms.VisualElement element, AndroidX.Fragment.App.FragmentManager fragmentManager, Android.Content.Context context) [0x00031] in D:\a\1\s\Xamarin.Forms.Platform.Android\Platform.cs:397 
  at Xamarin.Forms.Platform.Android.AppCompat.FragmentContainer.OnCreateView (Android.Views.LayoutInflater inflater, Android.Views.ViewGroup container, Android.OS.Bundle savedInstanceState) [0x00008] in D:\a\1\s\Xamarin.Forms.Platform.Android\AppCompat\FragmentContainer.cs:64 
  at AndroidX.Fragment.App.Fragment.n_OnCreateView_Landroid_view_LayoutInflater_Landroid_view_ViewGroup_Landroid_os_Bundle_ (System.IntPtr jnienv, System.IntPtr native__this, System.IntPtr native_inflater, System.IntPtr native_container, System.IntPtr native_savedInstanceState) [0x00021] in D:\a\1\s\generated\androidx.fragment.fragment\obj\Release\monoandroid90\generated\src\AndroidX.Fragment.App.Fragment.cs:1919 
  at (wrapper dynamic-method) Android.Runtime.DynamicMethodNameCounter.83(intptr,intptr,intptr,intptr,intptr)
```

**Comment by sthewissen** [comment]
Do you happen to implicitly or explicitly set a Visual="Material" to the view?

**Comment by matteopiccioni** [comment]
Yes, it crashes when I go into a ContentPage with Visual="Material" 
If I remove Visual="Material" the crash disappear

For now I have the stable version built directly from the source code, and in this way it works

Thanks

**Comment by sthewissen** [comment]
The newest beta might solve this (707) as it tries to set the `Visual` in the constructor to override the inheritance from the `ContentPage`.

**Comment by sthewissen** [comment]
Confirmed fixed in #116.
